sympy/polys/matrices/sdm.py

```diff
@@ -970,3 +1017,27 @@ def sdm_particular_from_rref(A, ncols, pivots):
         if Ain is not None:
             P[j] = Ain / A[i][j]
     return P
+
+
+def sdm_applyfunc(M, shape, f, domain):
+
+    fzero = f(domain.zero)
+    if not fzero:
+        return unop_dict(M, f)
+
+    nrows, ncols = shape
+
+    Mf = {i: {j: fzero for j in range(ncols)} for i in range(nrows)}
+
+    for i, Mi in M.items():
+        num_zero_elems = 0
+        for j, Mij in Mi.items():
+            fMij = f(Mij)
+            if fMij:
+                Mf[i][j] = fMij
+            else:
+                del Mf[i][j]
+                num_zero_elems += 1
+        if num_zero_elems == ncols:
+            del Mf[i]
```

Comment:
It would be better to do `Mfi = Mf[i]` once and then use `Mfi` for everything. We don't need `num_zero_elems` because we can just do `if Mfi`.

Reply:
ok